Microsoft.Toolkit.Uwp.UI/Extensions/DependencyObjectExtensions.cs

```diff
@@ -233,12 +558,127 @@ public static IEnumerable<DependencyObject> FindDescendants(this DependencyObjec
+        /// available <see cref="FindDescendantOrSelf{T}(DependencyObject)"/> overloads instead,
+        /// which will offer a more compact syntax as well as better performance in those cases.
+        /// </para>
+        /// </summary>
+        /// <param name="element">The root element.</param>
+        /// <param name="searchType">The search type to use to explore the visual tree.</param>
+        /// <returns>All the descendant <see cref="DependencyObject"/> instance from <paramref name="element"/>.</returns>
+        public static IEnumerable<DependencyObject> FindDescendantsOrSelf(this DependencyObject element, SearchType searchType)
+        {
+            // Depth-first traversal, with recursion
+            static IEnumerable<DependencyObject> FindDescendantsWithDepthFirstSearch(DependencyObject element)
+            {
+                yield return element;
+
+                int childrenCount = VisualTreeHelper.GetChildrenCount(element);
+
+                for (var i = 0; i < childrenCount; i++)
+                {
+                    DependencyObject child = VisualTreeHelper.GetChild(element, i);
+
+                    yield return child;
+
+                    foreach (DependencyObject childOfChild in FindDescendants(child))
+                    {
+                        yield return childOfChild;
+                    }
+                }
+            }
+
+            // Breadth-first traversal, with pooled local stack
+            static IEnumerable<DependencyObject> FindDescendantsWithBreadthFirstSearch(DependencyObject element)
+            {
+                yield return element;
+
+                using ArrayPoolBufferWriter<object> bufferWriter = ArrayPoolBufferWriter<object>.Create();
+
+                int childrenCount = VisualTreeHelper.GetChildrenCount(element);
+
+                for (int i = 0; i < childrenCount; i++)
+                {
+                    DependencyObject child = VisualTreeHelper.GetChild(element, i);
+
+                    yield return child;
 
-                foreach (DependencyObject childOfChild in FindDescendants(child))
+                    bufferWriter.Add(child);
+                }
+
+                for (int i = 0; i < bufferWriter.Count; i++)
                 {
-                    yield return childOfChild;
+                    DependencyObject parent = (DependencyObject)bufferWriter[i];
+
+                    childrenCount = VisualTreeHelper.GetChildrenCount(parent);
+
+                    for (int j = 0; j < childrenCount; j++)
+                    {
+                        DependencyObject child = VisualTreeHelper.GetChild(parent, j);
+
+                        yield return child;
+
+                        bufferWriter.Add(child);
+                    }
                 }
             }
+
+            static IEnumerable<DependencyObject> ThrowArgumentOutOfRangeExceptionForInvalidSearchType()
```

Comment:
Why make this a static method if the containing method is static, the content is one line and it's only used once?

Reply:
For clarity, same reason why eg. we mark private fields as private instead of just without any modifiers 😊

Reply:
I'm not sure that answers my question. We're using this throw once, locally, I don't see the point in wrapping a method call around it?

Reply:
Oh sorry, I didn't realize that's what you were confused about. The method is there because it enables the throw helper pattern, which results in better codegen versus just having the `throw` epxression inline. I wrote some detailed docs [here](https://docs.microsoft.com/en-us/windows/communitytoolkit/developer-tools/throwhelper) with an explanation and code sample if you're curious 🙂